### PR TITLE
list workspace with new group graph

### PR DESCRIPTION
**Pull request by dvoet**
https://broadinstitute.atlassian.net/browse/DSDEEPB-1801
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to develop 
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: Check configuration files in Jenkins in case they need changes
- [ ] **Submitter**: Verify swagger UI on dev server still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala [hunk 1]
```diff
@@ -78,22 +83,28 @@ class GraphAuthDAO extends AuthDAO with GraphDAO {
           .loop(
             "vtx", //start point of loop
             invert(isVertexOfClass(VertexSchema.Workspace)), //loop condition: stop walking the path once we've found a workspace node
-            isConnectedToTargetWorkspace(workspaceId)) //loop body: only emit paths to access maps connected to the target workspace
+            isVertexOfClass(VertexSchema.Workspace)) //loop body: emit paths to all workspaces
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala]: you don't need this if it's the negation of the loop condition

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala [hunk 2]
```diff
@@ -68,6 +68,11 @@ class GraphAuthDAO extends AuthDAO with GraphDAO {
...
 
   override def getMaximumAccessLevel(userSubjectId: String, workspaceId: String, txn: RawlsTransaction): WorkspaceAccessLevel = {
+    val workspaces = listWorkspaces(userSubjectId, txn)
+    workspaces.find(_.workspaceId == workspaceId).map(_.accessLevel).getOrElse(WorkspaceAccessLevels.NoAccess)
+  }
+
+  override def listWorkspaces(userSubjectId: String, txn: RawlsTransaction): Seq[WorkspacePermissionsPair] = {
     txn withGraph { db =>
 
       //Returns a list of paths starting with a group the user belongs in and ending with the workspace's access map.
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala]: update this comment

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala [hunk 3]
```diff
@@ -68,6 +68,11 @@ class GraphAuthDAO extends AuthDAO with GraphDAO {
   }
 
   override def getMaximumAccessLevel(userSubjectId: String, workspaceId: String, txn: RawlsTransaction): WorkspaceAccessLevel = {
+    val workspaces = listWorkspaces(userSubjectId, txn)
+    workspaces.find(_.workspaceId == workspaceId).map(_.accessLevel).getOrElse(WorkspaceAccessLevels.NoAccess)
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala]: :-1: overkill to get all the workspaces when you're only looking for one

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala [hunk 4]
```diff
@@ -78,22 +83,28 @@ class GraphAuthDAO extends AuthDAO with GraphDAO {
...
-        val accessLabel = accessMap.getEdges(Direction.OUT).filter( _.getVertex(Direction.IN) == group ).head.getLabel
-        val accessLevel = WorkspaceAccessLevels.withName(EdgeSchema.stripEdgeRelation(accessLabel))
-
-        WorkspaceAccessLevels.max(accessLevel, maxAccessLevel)
-      })
+      //The last 3 elements of each path are the group then the workspace access map then the workspace.
+      //The edge label between the group and the workspace access map corresponds to the access level for the group.
+      //It is possible for there to be more than 1 access level for a user for a workspace
+      //Group the paths by workspace then take the max access level for each workspace
+      val accessByWorkspace = accessMapPaths.toList.groupBy(path => path.last.asInstanceOf[Vertex].getProperty[String]("workspaceId"))
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala]: getProperty breaks encapsulation of the serialization code (maybe one day we'll base64encode our property names???) but i'm not sure i can recommend `loadObject[Workspace](vertex).workspaceId` as an alternative and keep my dignity

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala (hunk 3 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala]: I don't think there is an actual difference, all the gremlin pipelines execute locally (not on the orient server) so we are pulling back all those vertices anyway only to discard them.

Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala]: you're still doing all the work to build permissions lists for a bunch of workspaces you don't care about

Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala]: you could probably make us both happy by making the loop emit function a parameter or something.

**Comment by helgridly** [comment]
:on:

**Pull request merged**